index.js

```diff
@@ -102,6 +115,54 @@ function fastifyPostgres (fastify, options, next) {
     }
   }
 
+  fastify.addHook('onRoute', routeOptions => {
+    const useTransaction = routeOptions.useTransaction || (routeOptions.options && routeOptions.options.useTransaction)
+
+    if (useTransaction) {
+      // This will rollback the transaction if the handler fails at some point
+      const onError = async (req, reply, error) => {
+        req.transactionFailed = true
+
+        try {
+          await req.pg.query('ROLLBACK')
+        } catch (err) {
+          await req.pg.query('ROLLBACK')
+        }
+      }
+
+      routeOptions.onError = addHandler(routeOptions.onError, onError)
+    }
+
+    const preHandler = async (req, reply) => {
+      const client = await pool.connect()
+      req.pg = client
+
+      if (useTransaction) {
+        await req.pg.query('BEGIN')
+      }
+    }
+
+    // This will commit the transaction (or rollback if that fails) and also always
+    // release the client, regardless of error state or useTransaction value
+    const onSend = async (req, reply, payload) => {
+      try {
+        if (!req.transactionFailed && useTransaction) {
+          await req.pg.query('COMMIT')
+        }
+      } catch (err) {
+        if (useTransaction) {
```

Comment:
do we need this conditional? aren't we executing this code only if we already knew that we needed a transaction?